c2rust-analyze/tests/filecheck/fields.rs

```diff
@@ -0,0 +1,45 @@
+pub struct Data<'d> {
+    // final lifetime parameters: Data<'d, 'h0, 'h1, 'h2>
+
+    // 1: hypothetical pointer field lifetime 'h0 -> Data<'d, 'h0>
+    pub pi: *mut i32, // &'h0 i32
+
+    // 2: hypothetical pointer field lifetime 'h1 -> Data<'d, 'h0, 'h1>
+    // 3: A has no new parameters to bubble up
+    // 8: A has new lifetime param 'h2 -> Data<'d, 'h0, 'h1, 'h2>
+    pub pa: *mut A<'d>, // &'h1 A<'a, 'h0, 'h1, 'h2>
+
+    // 4: A has no new parameters to bubble up
+    pub a: A<'d>, // A<'a, 'h0, 'h1, 'h2>
+}
+
+pub struct A<'a> {
+    // final lifetime parameters: A<'a, 'h0, 'h1, 'h2>
+
+    // 5: Data has new lifetime params 'h0 and 'h1 -> A<'a, 'h0, 'h1>
+    // 9: Data has new lifetime param 'h2, but 'h2 is already in A
+    pub rd: &'a Data<'a>, // &'a Data<'a, 'h0, 'h1, 'h2>
+
+    // 6: hypothetical pointer field lifetime 'h2 -> A<'a, 'h0, 'h1, 'h2>
+    // 7: A has new lifetime params to bubble up, but they're already present
+    pub pra: *mut &'a mut A<'a>, // &'h2 &'a A<'a, 'h0, 'h1, 'h2>
+}
+
+// CHECK-LABEL: final labeling for "_field_access"
+// CHECK-DAG: ([[#@LINE+3]]: ppd): addr_of = UNIQUE, type = READ | WRITE | UNIQUE
+// CHECK-DAG: ([[#@LINE+2]]: ra): &mut A
+// CHECK-DAG: ([[#@LINE+1]]: ppd): &mut &mut Data
+unsafe fn _field_access<'d>(ra: &'d mut A<'d>, ppd: *mut *mut Data<'d>) {
+    // CHECK-DAG: ([[#@LINE+2]]: rd): addr_of = UNIQUE, type = READ | UNIQUE
+    // CHECK-DAG: ([[#@LINE+1]]: rd): &Data
+    let rd = (*(**ppd).a.pra).rd;
+
+    // CHECK-DAG: ([[#@LINE+2]]: pi): addr_of = UNIQUE, type = READ | WRITE | UNIQUE
+    // CHECK-DAG: ([[#@LINE+1]]: pi): &mut i32
+    let pi = rd.pi;
+    *pi = 3;
```

Comment:
@spernsteiner it appears that this would not be labeled as `&mut` unless it's explicitly written to, even though `Data::pi` is a `*mut` field. Is that expected? Is it because `c2rust` wouldn't translate something to `*mut` unless it was definitely written to?

Reply:
That's the expected behavior.  It only uses `&mut` for pointers with the `WRITE` permission, and pointers only get `WRITE` permission if the pointer or something derived from it is written to.  Nothing in the analysis pays attention to whether the original C pointer was `const` or not (a lot of C code is not const-correct - it's common to just use `foo*` everywhere, even when `const foo*` would make more sense).